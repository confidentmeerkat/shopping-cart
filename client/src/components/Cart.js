import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import axios from 'axios';
import numeral from 'numeral';
import { getCart } from '../actions/cartActions';
import CheckoutModal from './CheckoutModal';
import FlatButton from 'material-ui/FlatButton';
import RaisedButton from 'material-ui/RaisedButton';
import NavigateNext from 'material-ui/svg-icons/image/navigate-next';
import RemoveShoppingCart from 'material-ui/svg-icons/action/remove-shopping-cart.js';
import Dialog from 'material-ui/Dialog';
import Snackbar from 'material-ui/Snackbar';
import '../styles/Cart.css';

class Cart extends React.Component {

  state = {
    checkoutModalOpen: false,
    dialogOpen: false,
    snackbarOpen: false
  }

  toggleCheckoutModal = () => {
    this.setState({ checkoutModalOpen: !this.state.checkoutModalOpen });
  }

  toggleDialog = (value) => {
    this.setState({ dialogOpen: value });
  }

  removeItem = (itemId) => {
    axios.put('/api/cart', {
        cartId: this.props.cart._id,
        itemId: itemId
      })
      .then(() => {
        this.props.getCart();
        this.setState({ snackbarOpen: true });
      });
  }

  emptyCart = () => {
    axios.delete('/api/cart', { params: { id: this.props.cart._id } })
      .then(() => {
        this.toggleDialog(false);
        this.props.getCart();
      });
  }

  makeOrder = () => {
    const order = this.props.cart.items.map((item) => {
      let order = {
        name: item.product.info.name,
        price: item.product.info.price,
        quantity: item.quantity,
        dateCreated: Date.now()
      };
      return order;
    });

    axios.post('/api/user/order', { order: order })
      .then(() => {
        this.toggleCheckoutModal();
        this.emptyCart();
      });
  }

  componentDidMount() {
    this.props.getCart();
  }

  render() {
    return (
      <div>
        <h1>Your Cart</h1>
        <div className="cart">
          <div className="cart-info">
            <p>
              <b>Number of items: </b>
              {this.props.cart.items.length ? this.props.cart.items.reduce((acc, item) => acc += item.quantity, 0) : 0}
            </p>
            <p>
              <b>Total amount: </b>
              <span className="total">
                {this.props.cart.items.length ? numeral(this.props.cart.items.reduce((acc, item) => acc += item.product.info.price * item.quantity, 0)).format('$0,0.00') : numeral(0).format('$0,0.00')}
              </span>
            </p>
            <div className="btns">
              <RaisedButton
                onClick={this.toggleCheckoutModal}
                className="btn"
                label="Checkout"
                labelPosition="before"
                icon={<NavigateNext />}
                primary={true}
                disabled={!this.props.cart.items.length}
              />
              <RaisedButton
                onClick={() => this.toggleDialog(true)}
                className="btn"
                label="Empty cart"
                labelPosition="before"
                icon={<RemoveShoppingCart />}
                secondary={true}
                disabled={!this.props.cart.items.length}
              />
            </div>
            <CheckoutModal
              isOpen={this.state.checkoutModalOpen}
              onRequestClose={this.toggleCheckoutModal}
              toggle={this.toggleCheckoutModal}
              makeOrder={this.makeOrder}           
            />
            <Dialog
              title="Are you sure that you want to empty your cart?"
              actions={[
                <FlatButton
                  label="Cancel"
                  primary={true}
                  onClick={() => this.toggleDialog(false)}
                />,
                <FlatButton
                  label="Yes"
                  primary={true}
                  onClick={this.emptyCart}
                />,
              ]}
              modal={true}
              open={this.state.dialogOpen}
            >
              All items will be removed.
            </Dialog>
          </div>
          <div className="cart-items">
            <table>
              <tr>
                <th></th>
                <th>Product Name</th>
                <th>Price</th>
                <th>Quantity</th>
                <th>Total</th>
                <th></th>
              </tr>
              {this.props.cart.items.length ? this.props.cart.items.map((item) => {
                return (
                  <tr>
                    <td><img src={item.product.info.photo} /></td>
                    <td><Link to={`/product/${item.product._id}`}>{item.product.info.name}</Link></td>
                    <td>{numeral(item.product.info.price).format('$0,0.00')}</td>
                    <td>{item.quantity}</td>
                    <td>{numeral(item.product.info.price * item.quantity).format('$0,0.00')}</td>
                    <td><button title="Remove this item from the cart" onClick={() => this.removeItem(item._id)}>X</button></td>
                  </tr>
                );
              }) : ''}
            </table>
            {!this.props.cart.items.length && <h1>No items in the cart.</h1>}
            <Snackbar
              open={this.state.snackbarOpen}
              message="Item removed from your cart."
              autoHideDuration={4000}
              bodyStyle={{ 'textAlign': 'center' }}
            />
          </div>
        </div>
      </div>
    )
  }
};

const mapStateToProps = (state) => ({
  cart: state.cart
})

const mapDispatchToProps = (dispatch) => ({
  getCart: () => dispatch(getCart())
});

export default connect(mapStateToProps, mapDispatchToProps)(Cart);