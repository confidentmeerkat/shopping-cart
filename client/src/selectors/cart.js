export const selectCart = (state) => state.cart;
export const selectItems = (state) => state.cart.items;