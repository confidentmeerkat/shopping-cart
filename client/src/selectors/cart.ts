import { IState, ICart, ICartProduct } from '@typings/state/index';

export const selectCart = (state: IState): ICart => state.cart;
export const selectItems = (state: IState): ICartProduct[] => state.cart.items;
