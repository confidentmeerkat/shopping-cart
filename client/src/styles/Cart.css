.cart-container {
  min-height: 100vh;
  margin-top: -72px;
  border: 1px solid #ffffff00;
}

.cart-container h1 {
  margin-top: 100px;
  margin-bottom: 40px;
}

.cart {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

.cart-info {
  display: flex;
  flex-direction: column;
  width: 20%;
  height:100%;
  min-height: auto;
  background: #f5f5f5;
}

.cart-info p {
  padding: 0 10px;
}

.cart-info .total {
  font-size: 22px;
  color: #64DD17;
}

.cart-info .btns {
  text-align: center;
}

.cart-info .btn {
  width: 150px;
  margin: 10px auto;
}

.cart-items {
  width: 78%;
}

.cart-items img {
  max-width: 50px;
}

.cart-items table {
  width: 100%;
  border-spacing: 0;
}

.cart-items th {
  padding: 5px 5px;
  background-color: #00BCD4;
  color: white;
  text-align: left;
}

.cart-items th:first-child {
  width: 70px;
}

.cart-items th:last-child {
  width: 15px;
}

.cart-items td {
  padding: 12px 5px;
  text-align: left;
  border-bottom: 1px solid #00BCD4;
}

.cart-items td:nth-child(4) {
  text-align: center;
}

.cart-items a {
  color: inherit;
  text-decoration: none;
}

.cart-items button {
  padding: 1px 5px;
  background: none;
  color: #F44336;
  font-weight: bold;
  border: none;
  border-radius: 50%;
  cursor: pointer;
}

.cart-items button:hover {
  background: #F44336;
  color: white;
}

.cart-items h1 {
  text-align: center;
  margin-top: 100px;
}

@media (max-width: 1000px) {
  .cart {
    flex-direction: column;
  }

  .cart-info {
    flex-direction: row;
    width: 100%;
  }

  .cart-info .info {
    width: 50%;
  }

  .cart-info .btns {
    width: 50%;
  }

  .cart-info .btn {
    margin: 10px;
  }

  .cart-items {
    width: 100%;
    margin: 40px auto 60px auto;
  }
}