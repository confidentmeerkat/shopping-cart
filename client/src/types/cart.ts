import { Product } from './product';

interface Item {
  _id: string;
  product: Product;
  quantity: number;
}

export interface Cart {
  isLoading: boolean;
  isLoaded: boolean;
  _id: string;
  items: Item[];
  error: string | null;
}
