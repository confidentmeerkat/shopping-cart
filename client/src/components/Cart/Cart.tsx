import * as React from 'react';
import { Link } from 'react-router-dom';
import axios from 'axios';
import * as numeral from 'numeral';
import FlatButton from 'material-ui/FlatButton';
import RaisedButton from 'material-ui/RaisedButton';
import NavigateNext from 'material-ui/svg-icons/image/navigate-next';
import RemoveShoppingCart from '@material-ui/icons/RemoveShoppingCart';
import Dialog from 'material-ui/Dialog';
import Snackbar from 'material-ui/Snackbar';
import CheckoutModal from '../CheckoutModal';
import OrderSuccessModal from '../OrderSuccessModal';
import { ICart } from '@typings/state/index';
import { modal } from '@typings/modal';
import '@styles/Cart.css';

interface Props {
  cart: ICart;
  getCart: () => ICart;
}

interface State {
  activeModal: modal
}

class Cart extends React.Component<Props, State> {
  state = {
    activeModal: null
  }

  setActiveModal = (modal: modal) => {
    this.setState({ activeModal: modal });
  }

  removeItem = async (itemId: string) => {
    await axios.put('/api/cart', {
      cartId: this.props.cart._id,
      itemId: itemId
    });

    this.props.getCart();

    this.setActiveModal('snackbar');
    setTimeout(() => {
      this.setActiveModal(null);
    }, 4000);
  }

  emptyCart = async () => {
    await axios.delete('/api/cart', { params: { id: this.props.cart._id } });
    await this.setState({ activeModal: null });
    await this.props.getCart();
  }

  makeOrder = async () => {
    const order = this.props.cart.items.map((item) => {
      let order = {
        name: item.product.info.name,
        price: item.product.info.price,
        quantity: item.quantity,
        dateCreated: Date.now()
      };
      return order;
    });

    await axios.post('/api/order', { order: order });
    await this.emptyCart();
    
    this.setActiveModal('orderSuccess');
  }

  componentWillMount() {
    this.props.getCart();
  }

  render() {
    const { cart } = this.props;
    const cartExists = cart.isLoaded && !cart.error && cart.items.length;

    return (
      <div className="cart-container">
        <h1>Your Cart</h1>
        <div className="cart">
          <div className="cart-info">
            <div className="info">
              <p>
                <b>Number of items: </b>
                {cartExists ? cart.items.reduce((acc, item) => acc += item.quantity!, 0) : 0}
              </p>
              <p>
                <b>Total amount: </b>
                <span className="total">
                  {cartExists ? numeral(cart.items.reduce((acc, item) => acc += item.product.info.price * item.quantity!, 0)).format('$0,0.00') : numeral(0).format('$0,0.00')}
                </span>
              </p>
            </div>
            <div className="btns">
              <RaisedButton
                onClick={() => this.setActiveModal('checkout')}
                className="btn"
                label="Checkout"
                labelPosition="before"
                icon={<NavigateNext />}
                primary={true}
                disabled={!cartExists}
              />
              <RaisedButton
                onClick={() => this.setActiveModal('dialog')}
                className="btn"
                label="Empty cart"
                labelPosition="before"
                icon={<RemoveShoppingCart />}
                secondary={true}
                disabled={!cartExists}
              />
            </div>
            <CheckoutModal
              isOpen={this.state.activeModal === 'checkout'}
              onRequestClose={() => this.setActiveModal}
              setActiveModal={this.setActiveModal}
              makeOrder={this.makeOrder}
            />
            <OrderSuccessModal
              isOpen={this.state.activeModal === 'orderSuccess'}
              setActiveModal={this.setActiveModal}
            />
            <Dialog
              title="Are you sure that you want to empty your cart?"
              actions={[
                <FlatButton
                  label="Cancel"
                  primary={true}
                  onClick={() => this.setActiveModal(null)}
                />,
                <FlatButton
                  label="Yes"
                  primary={true}
                  onClick={this.emptyCart}
                />,
              ]}
              modal={true}
              open={this.state.activeModal === 'dialog'}
            >
              All items will be removed.
            </Dialog>
          </div>
          <div className="cart-items">
            {cartExists ?
              <table>
                <thead>
                  <tr>
                    <th></th>
                    <th>Product Name</th>
                    <th>Price</th>
                    <th>Qty</th>
                    <th>Total</th>
                    <th></th>
                  </tr>
                </thead>
                <tbody>
                  {cart.items.map((item) => {
                    return (
                      <tr key={item.product.info.name} >
                        <td><img src={item.product.info.photo} /></td>
                        <td><Link to={`/product/${item.product._id}`}>{item.product.info.name}</Link></td>
                        <td>{numeral(item.product.info.price).format('$0,0.00')}</td>
                        <td>{item.quantity}</td>
                        <td>{numeral(item.product.info.price * item.quantity!).format('$0,0.00')}</td>
                        <td><button title="Remove this item from the cart" onClick={() => this.removeItem(item._id)}>X</button></td>
                      </tr>
                    );
                  })}
                </tbody>
              </table> :
              <h1>No items in the cart.</h1>
            }
            <Snackbar
              open={this.state.activeModal === 'snackbar'}
              message="Item removed from your cart."
              bodyStyle={{ 'textAlign': 'center' }}
            />
          </div>
        </div>
      </div>
    )
  }
};

export default Cart;
