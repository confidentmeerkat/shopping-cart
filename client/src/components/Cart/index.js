import { compose } from 'recompose';
import { connect } from 'react-redux';
import { getCart } from '@actions';
import { selectCart } from '@selectors/cart';
import Cart from './Cart';

const mapStateToProps = (state) => ({
  cart: selectCart(state)
});

const actions = { getCart };

export default compose(
  connect(mapStateToProps, actions)
)(Cart);
