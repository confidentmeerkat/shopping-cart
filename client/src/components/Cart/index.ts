import { compose } from 'recompose';
import { connect } from 'react-redux';
import { getCart } from '@actions/index';
import { selectCart } from '@selectors/cart';
import { IState } from '@typings/state/index';
import Cart from './Cart';

const mapStateToProps = (state: IState) => ({
  cart: selectCart(state)
});

const actions = { getCart };

export default compose(
  connect(mapStateToProps, actions)
)(Cart);
