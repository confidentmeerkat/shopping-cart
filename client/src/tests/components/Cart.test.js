import React from 'react';
import { shallow } from 'enzyme';
import { Cart } from '../../components/Cart';
import cart from '../__mocks__/cart';

describe('<Cart />', () => {
  it('renders Cart correctly', () => {
    const wrapper = shallow(<Cart cart={cart} getCart={() => { }} />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.info').childAt(0).text()).toBe('Number of items: 3');
    expect(wrapper.find('.info').childAt(1).text()).toBe('Total amount: $2,200.00');
    expect(wrapper.find('tbody').children()).toHaveLength(2);
    expect(wrapper.find('tbody').childAt(0).children()).toHaveLength(6);
  });

  it('triggers right action when clicked on buttons', () => {
    const wrapper = shallow(<Cart cart={cart} getCart={() => { }} />);

    wrapper.find('RaisedButton').at(0).simulate('click');
    expect(wrapper.state('checkoutModalOpen')).toBe(true);
  });
});