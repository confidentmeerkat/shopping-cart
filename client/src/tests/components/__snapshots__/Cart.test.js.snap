// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> renders Cart correctly 1`] = `
<div
  className="cart-container"
>
  <h1>
    Your Cart
  </h1>
  <div
    className="cart"
  >
    <div
      className="cart-info"
    >
      <div
        className="info"
      >
        <p>
          <b>
            Number of items: 
          </b>
          3
        </p>
        <p>
          <b>
            Total amount: 
          </b>
          <span
            className="total"
          >
            $2,200.00
          </span>
        </p>
      </div>
      <div
        className="btns"
      >
        <RaisedButton
          className="btn"
          disabled={false}
          fullWidth={false}
          icon={<ImageNavigateNext />}
          label="Checkout"
          labelPosition="before"
          onClick={[Function]}
          primary={true}
          secondary={false}
        />
        <RaisedButton
          className="btn"
          disabled={false}
          fullWidth={false}
          icon={<ActionRemoveShoppingCart />}
          label="Empty cart"
          labelPosition="before"
          onClick={[Function]}
          primary={false}
          secondary={true}
        />
      </div>
      <Connect(CheckoutModal)
        isOpen={false}
        makeOrder={[Function]}
        onRequestClose={[Function]}
        toggle={[Function]}
      />
      <OrderSuccessModal
        isOpen={false}
        onRequestClose={[Function]}
        toggle={[Function]}
      />
      <Dialog
        actions={
          Array [
            <FlatButton
              disabled={false}
              fullWidth={false}
              label="Cancel"
              labelPosition="after"
              labelStyle={Object {}}
              onClick={[Function]}
              onKeyboardFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onTouchStart={[Function]}
              primary={true}
              secondary={false}
            />,
            <FlatButton
              disabled={false}
              fullWidth={false}
              label="Yes"
              labelPosition="after"
              labelStyle={Object {}}
              onClick={[Function]}
              onKeyboardFocus={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onTouchStart={[Function]}
              primary={true}
              secondary={false}
            />,
          ]
        }
        autoDetectWindowHeight={true}
        autoScrollBodyContent={false}
        modal={true}
        open={false}
        repositionOnUpdate={true}
        title="Are you sure that you want to empty your cart?"
      >
        All items will be removed.
      </Dialog>
    </div>
    <div
      className="cart-items"
    >
      <table>
        <thead>
          <tr>
            <th />
            <th>
              Product Name
            </th>
            <th>
              Price
            </th>
            <th>
              Qty
            </th>
            <th>
              Total
            </th>
            <th />
          </tr>
        </thead>
        <tbody>
          <tr
            key="Apple iPhone"
          >
            <td>
              <img />
            </td>
            <td>
              <Link
                replace={false}
                to="/product/undefined"
              >
                Apple iPhone
              </Link>
            </td>
            <td>
              $700.00
            </td>
            <td>
              1
            </td>
            <td>
              $700.00
            </td>
            <td>
              <button
                onClick={[Function]}
                title="Remove this item from the cart"
              >
                X
              </button>
            </td>
          </tr>
          <tr
            key="Samsung Galaxy S8"
          >
            <td>
              <img />
            </td>
            <td>
              <Link
                replace={false}
                to="/product/undefined"
              >
                Samsung Galaxy S8
              </Link>
            </td>
            <td>
              $750.00
            </td>
            <td>
              2
            </td>
            <td>
              $1,500.00
            </td>
            <td>
              <button
                onClick={[Function]}
                title="Remove this item from the cart"
              >
                X
              </button>
            </td>
          </tr>
        </tbody>
      </table>
      <Snackbar
        bodyStyle={
          Object {
            "textAlign": "center",
          }
        }
        message="Item removed from your cart."
        open={false}
      />
    </div>
  </div>
</div>
`;
